# U.S. Medical Insurance Project

For this project, I will be investigating a medical insurance costs dataset in a .csv file using the Python skills that I have developed during the Codecademy Data Science career. 

First, we took a look to the csv file in order to know what information we have available.

As we can see, there are seven categories that we can use for this project. These fields are about their age, sex, bmi, the number of children that they have, whether they are smoker or not, the region where they are from and the result of all of this columns that are the charges.
So, in our python file, we start by creating seven empty lists that are releated to our fields.

In [148]:
ages = list()
sexes = list()
bmis = list()
numOfChildren = list()
isSmoker = list()
regions = list()
charges = list()

Then, since we have a csv file and we want to work with it, we have to get and distribute all the data.

In [151]:
# We are working with a csv file, so we need to import the csv library
import csv

# We create a function that we are gonna use soon in order to put the data of the csv into the lists.
def getData(listName, column, CSVFile="insurance.csv"):
    data = dict()
    
    with open(CSVFile) as file:
        data = csv.DictReader(file, delimiter=",")

        for row in data:
            listName.append(row[column])

After we create our function that get the csv information, we have to put them into the lists so we call the function seven times.

In [154]:
getData(ages, "age")
getData(sexes, "sex")
getData(bmis, "bmi")
getData(numOfChildren, "children")
getData(isSmoker, "smoker")
getData(regions, "region")
getData(charges, "charges")

Now that we have all the information in their respective list we would create a python class where we are going to work with some methods so we would be able to analize the data that we got.

In [157]:
class analizeData:
    # First, we create the constructor with the seven parameters.
    def __init__(self, allAges, allSexes, allBmis, allChildren, allSmoker, allRegions, allCharges):
        self.allAges = allAges
        self.allSexes = allSexes
        self.allBmis = allBmis
        self.allChildren = allChildren
        self.allSmoker = allSmoker
        self.allRegions = allRegions
        self.allCharges = allCharges

        self.putAllIntoDictionaries()

    # Second, we want all the information together so we would only need to work with a big list with a dictionary that agrupates the seven respective values.
    def putAllIntoDictionaries(self):
        self.listOfDictionaries = list()
        
        for i in range(len(self.allAges)):
            dictionary = dict()
            dictionary["age"] = self.allAges[i]
            dictionary["sex"] = self.allSexes[i]
            dictionary["bmi"] = self.allBmis[i]
            dictionary["children"] = self.allChildren[i]
            dictionary["smoker"] = self.allSmoker[i]
            dictionary["region"] = self.allRegions[i]
            dictionary["charges"] = self.allCharges[i]
            self.listOfDictionaries.append(dictionary)

    # Third, we want to order all the information that we got by ordering by age.
    def segregateChargesByAge(self):
        lista = list()
        newList = list()

        # In this iteration we put into a list a tuple that has the age and charges camp of each dictionary.
        for dictionary in self.listOfDictionaries:
                lista.append((dictionary["age"], dictionary["charges"]))

        # Then, we sort them, so we would be able to have the information organized.
        lista.sort()

        for element in lista:
            # In this nested loop we select the dictionary that complies with the both camps that we have mentioned before, and as we have the previous
            # list sorted we would be able to pick each dictionary in order
            for dictionary in self.listOfDictionaries:
                if element[0] == dictionary["age"] and element[1] == dictionary["charges"]:
                    newList.append(dictionary)

        for element in newList:
            print(element)

    # Fourth, we want to do the same as the third part but now, first, we want to order the data by region.
    def agrupationByRegionThenByAgeThenByCharges(self):
        lista = list()

        # So, in a list with include a tuple that contains region, age and charges fields.
        for dictionary in self.listOfDictionaries:
            lista.append((dictionary["region"], dictionary["age"], dictionary["charges"]))

        lista.sort()

        # Now, we create four empty lists that are separated by the regions that we know.
        listNE = list()
        listNW = list()
        listSE = list()
        listSW = list()

        for element in lista:
            # For this part, we have to see if the camps of the dictionary have the same atributes that we want for each region.
            for dictionary in self.listOfDictionaries:
                if element[0] == dictionary["region"] == "northeast" and element[1] == dictionary["age"] and element[2] == dictionary["charges"]:
                    listNE.append(dictionary)

                elif element[0] == dictionary["region"] == "northwest" and element[1] == dictionary["age"] and element[2] == dictionary["charges"]:
                    listNW.append(dictionary)
                
                elif element[0] == dictionary["region"] == "southeast" and element[1] == dictionary["age"] and element[2] == dictionary["charges"]:
                    listSE.append(dictionary)
                
                elif element[0] == dictionary["region"] == "southwest" and element[1] == dictionary["age"] and element[2] == dictionary["charges"]:
                    listSW.append(dictionary)

        # Now, we put the region lists into a dictionary in order to get all the information together.
        dictionary = {"northeast": listNE, "northwest": listNW, "southeast": listSE, "southwest": listSW}

        for key, value in dictionary.items():
            print(key.upper())

            for element in value:
                print(element)

            print("\n")               

    # Fifth, we want to analize the age by finding the main, median and mode.
    def analizingAge(self):
        self.allAges.sort()
        length = len(self.allAges)

        # In this case the age data is an string so we tranform it into a string.
        for i in range(length):
            self.allAges[i] = int(self.allAges[i])

        # The main is calculated by the division of the sum and average of the integer list.
        main = sum(self.allAges) / length
        
        median = -1
        mode = [[0], 0]

        # When we want to find the median we have to cases.

        # The first one is if the length is an even number, so we have to find the two numbers that are in the middle and calculate the semi-sum.
        if length % 2 == 0:
            median = (self.allAges[int(length/2) - 1] + self.allAges[int(length/2)]) / 2

        # The second one is only picking the middle number of the list, since the length is odd.
        else:
            median = self.allAges[int(length/2)]

        # For the mode case, we have to count how much a number repeats and if it's bigger than the previous one we select it. Remember that we can have
        # more than one mode, so that's why we have a sublist inside the list of mode.
        
        for age in self.allAges:
            if self.allAges.count(age) >= mode[1]:
                cont = 0
                for ages in mode[0]:
                    if age == ages:
                        cont += 1

                if cont == 0:
                    mode[0].append(age)


                mode[1] = self.allAges.count(age)

        mode[0].remove(0)
        
        print("Main: " + str(main))
        print("Median: " + str(median))
        
        print("Mode: ")
        print(mode[0])
        print("times repeating: " + str(mode[1]))

    # This is a function that is used in the sixth part that consist in rounding a number manually, not by using the function round().
    def privateRound(self, stringNum):
        string = ""
        decimal = False
        position = 0
        length = len(stringNum)

        # We make a loop in order to know if it has decimal point or not. If so, we would find its position.
        for char in stringNum:
            if char == ".":
                decimal = True
                break

            else:
                string += char
                position += 1

        # After the loop, we calculate the difference between the length of the string and the position of the position so we would know how many decimals
        # it has.
        dif = length - position - 1

        # Then, we create a dictionary with six keys that represents each decimal, why 6? For a better rounding.
        values = {"sixthDecimal": [5, 0],
                  "fifthDecimal": [4, 0],
                  "fourthDecimal": [3, 0],
                  "thirdDecimal": [2, 0],
                  "secondDecimal": [1, 0],
                  "firstDecimal": [0, 0]}

        # And we set a boolean carry so we would know if the next decimal will need to increase by one.
        carry = False

        # In this loop we start to see if there exist decimals for the number. If so, we check how many. 
        for value in values.values():

            # If the difference is bigger than the position of the decimal, we would update the second value of the list of the dicitonary used value.
            if dif > value[0]:
                value[1] = int(stringNum[position + value[0] + 1])

                # Then, inside that if statement, we chek if there was a previous carry. If so, we add one to that decimal number.
                if carry: value[1] += 1

                # Then, we check whether the decimal number is bigger than four. If so, we set the carry value as True. If not, as False.
                if value[1] > 4: carry = True
                else: carry = False

        # After the loop we transform the string value into an integer.
        stringNum = int(string)

        # Then, if the carry was set on True, we add one to that integer.
        if carry:
            stringNum += 1

        return stringNum   

    # Sixth, we analize the charges by finding its minimum, maximum and average value.    
    def analizingCharges(self):
        
        pos = 0        
        # We have the charges information as an string object, so we tranform them into integers with the previos function.
        for charge in self.allCharges:
            self.allCharges[pos] = self.privateRound(charge)
            pos += 1            

        # Now, we find each value, then we print them.
        minimum = min(self.allCharges)
        maximum = max(self.allCharges)
        main = sum(self.allCharges) / len(self.allCharges)
        main = self.privateRound(str(main))

        print("Minimum charge: " + str(minimum))
        print("Maximum charge: " + str(maximum))
        print("Main charge: " + str(main))
        
        
        

After making the class, we create its respective object.

In [160]:
allData = analizeData(ages, sexes, bmis, numOfChildren, isSmoker, regions, charges)

First, we want to check how the charges are distributed by age.

In [163]:
allData.segregateChargesByAge()

{'age': '18', 'sex': 'male', 'bmi': '23.21', 'children': '0', 'smoker': 'no', 'region': 'southeast', 'charges': '1121.8739'}
{'age': '18', 'sex': 'male', 'bmi': '28.31', 'children': '1', 'smoker': 'no', 'region': 'northeast', 'charges': '11272.33139'}
{'age': '18', 'sex': 'male', 'bmi': '30.14', 'children': '0', 'smoker': 'no', 'region': 'southeast', 'charges': '1131.5066'}
{'age': '18', 'sex': 'male', 'bmi': '33.33', 'children': '0', 'smoker': 'no', 'region': 'southeast', 'charges': '1135.9407'}
{'age': '18', 'sex': 'male', 'bmi': '33.66', 'children': '0', 'smoker': 'no', 'region': 'southeast', 'charges': '1136.3994'}
{'age': '18', 'sex': 'male', 'bmi': '34.1', 'children': '0', 'smoker': 'no', 'region': 'southeast', 'charges': '1137.011'}
{'age': '18', 'sex': 'male', 'bmi': '34.43', 'children': '0', 'smoker': 'no', 'region': 'southeast', 'charges': '1137.4697'}
{'age': '18', 'sex': 'male', 'bmi': '37.29', 'children': '0', 'smoker': 'no', 'region': 'southeast', 'charges': '1141.4451'}


Second, we want to know how the previous information is distributed first by the region.

In [165]:
allData.agrupationByRegionThenByAgeThenByCharges()

NORTHEAST
{'age': '18', 'sex': 'male', 'bmi': '28.31', 'children': '1', 'smoker': 'no', 'region': 'northeast', 'charges': '11272.33139'}
{'age': '18', 'sex': 'male', 'bmi': '17.29', 'children': '2', 'smoker': 'yes', 'region': 'northeast', 'charges': '12829.4551'}
{'age': '18', 'sex': 'male', 'bmi': '39.14', 'children': '0', 'smoker': 'no', 'region': 'northeast', 'charges': '12890.05765'}
{'age': '18', 'sex': 'male', 'bmi': '21.565', 'children': '0', 'smoker': 'yes', 'region': 'northeast', 'charges': '13747.87235'}
{'age': '18', 'sex': 'female', 'bmi': '21.66', 'children': '0', 'smoker': 'yes', 'region': 'northeast', 'charges': '14283.4594'}
{'age': '18', 'sex': 'male', 'bmi': '25.175', 'children': '0', 'smoker': 'yes', 'region': 'northeast', 'charges': '15518.18025'}
{'age': '18', 'sex': 'male', 'bmi': '15.96', 'children': '0', 'smoker': 'no', 'region': 'northeast', 'charges': '1694.7964'}
{'age': '18', 'sex': 'male', 'bmi': '21.47', 'children': '0', 'smoker': 'no', 'region': 'northeas

Third, we find some important values of the age

In [167]:
allData.analizingAge()

Main: 39.20702541106129
Median: 39.0
Mode: 
[18]
times repeating: 69


We have found that the average age is 39 aproximately. Than the value in the center, if we order them,  is 39 too.
And the age that is repeated the most is 18 by 69 times.

Fourth, we check the charges values

In [171]:
allData.analizingCharges()

Minimum charge: 1122
Maximum charge: 63770
Main charge: 13271


As integer values, aproximated by the "privateRounding()" function, we have gotten that the minimum charge is 1122 dollars. The maximum, 63770 dollars. And the average is 13271 dollars aproximally.